README.md
<!--

WARNING: This file is auto-generated from ./update-readme.sh, modify
that instead.

-->

# ghc-musl

This repository provides Docker images with GHC compiled with `musl`;
therefore can be used to create fully static Haskell binaries without
`glibc` dependency on any platform which can run Docker (x86_64). Powered
by [ghcup](https://www.haskell.org/ghcup/).

Images come with `ghc` and `cabal` executables alongside with commonly
used libraries and build tools. They can also be used with the `stack`
build tool using its Docker integration.

Here are the latest images currently [published in Docker Hub](https://hub.docker.com/r/utdemir/ghc-musl/tags):

* utdemir/ghc-musl:v25-ghc944
* utdemir/ghc-musl:v25-ghc925
* utdemir/ghc-musl:v25-ghc902
* utdemir/ghc-musl:v25-ghc8107
* utdemir/ghc-musl:v25-ghc884

## Usage

### cabal-install

Mount the project directory to the container, and use `cabal-install`
with `--enable-executable-static` flag inside the container:

```
$ cd myproject/
$ docker run -itv $PWD:/mnt utdemir/ghc-musl:v25-ghc944
sh$ cd /mnt
sh$ cabal new-update
sh$ cabal new-build --enable-executable-static
```

You can also set `executable-static` [option](https://cabal.readthedocs.io/en/latest/cabal-project.html#cfg-field-executable-static) on your `cabal.project` file.

### stack

Use below arguments, or set the relevant options on your `stack.yaml`:

```
stack build \
  --ghc-options ' -static -optl-static -optl-pthread -fPIC' \
  --docker --docker-image "utdemir/ghc-musl:v25-ghc944" \
  --no-nix
```

Make sure to pick an image with the GHC version compatible with the
Stackage resolver you are using.

Follow https://github.com/commercialhaskell/stack/issues/3420 for
more details on static compilation using the Stack build tool.

### Example session with GHC

Below shell session shows how to start a pre-compiled docker container
and compile a simple `Hello.hs` as a static executable:

```
$ docker run -itv $PWD:/mnt utdemir/ghc-musl:v25-ghc944
bash-4.4# cd /mnt/
bash-4.4# cat Hello.hs
main = putStrLn "Hello"
bash-4.4# ghc --make -optl-static -optl-pthread Hello.hs
[1 of 1] Compiling Main             ( Hello.hs, Hello.o )
Linking Hello ...
bash-4.4# ls -al Hello
-rwxr-xr-x 1 root root 1185056 Aug 11 16:55 Hello
bash-4.4# file Hello
Hello: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), statically linked, with debug_info, not stripped
bash-4.4# ldd ./Hello
ldd: ./Hello: Not a valid dynamic program
```

The result can be executed on different Linux systems:

```
$ docker run -itv $PWD:/mnt alpine
# /mnt/Hello
Hello
```

```
$ docker run -itv $PWD:/mnt centos
[root@94eb29dcdfe6 /]# /mnt/Hello
Hello
```

## Development

Feel free to open an issue or send a PR to add a library or support a
newer compiler version.

The build process is orchestrated using [Earthly][]. Once it is installed
(or obtained via "nix-shell -p earthly"), you can use below command to build
and test all images:

```
earthly --allow-privileged +all
```

Note: `--allow-privileged` is only necessary because we use Docker-in-Docker
project to test if the generated images work with `stack`'s Docker support.
You can use the following command to disable Stack tests so that
`--allow-privileged` is not necessary:

```
earthly --build-arg TEST_STACK=0 +all
```

The following command updates `README.md`:

```
earthly --artifact +readme/README.md
```

[Earthly]: https://earthly.dev

## Related

* <https://gitlab.com/neosimsim/docker-builder-images>
* Using Nix: <https://github.com/nh2/static-haskell-nix>
* Not maintained: <https://github.com/fpco/docker-static-haskell>


CHANGELOG.md
# v25

* ghc: 9.2.2 -> 9.2.5
* ghc: 9.4.4
* alpine: 3.17.1

# v24

* ghc: 9.2.1 -> 9.2.2 (thanks to @TravisCardwell)
* ghc: 8.8.4 added (thanks to @TravisCardwell)
* alpine: 3.14.0 -> 3.15.0

# v23

* ghc: 9.0.1 -> 9.0.2

# v22

* ghc: 9.2.1
* ghc: 8.8.4 removed

# v21

* ghc: 8.10.4 -> 8.10.7
* alpine: 3.14.0 -> 3.14.2

# v20

* New library: freetype, libpng, brotli
* Update Alpine version from 3.13 to 3.14
* Bugfix: symlink libncurses as libtinfo

# v19

* New libraries: sdl, sdl2

# v18

* Update cabal: 3.4-rc1 -> 3.4

# v17

* New compiler: GHC 9.0.1 with cabal 3.4
* Updated Alpine version from 3.12 to 3.13
* Removed compiler: GHC 8.6.5

# v16

* New executables: groupadd, useradd

# v15

* Install GHC under /usr/local instead of /root/
* GHC 8.10.1 -> 8.10.2

# v14

* Build system revamp, now based on Alpine.

# v13

* New libraries: bzip2, openssl

# v12

* New libraries: pcre, pcre2
* New executables: tar, dpkg, fakeroot

# v11

* New executables: find, xargs, perl, gzip, file

# v10

* bugfix: make sure that curl can find SSL certificates

# v9

* New libraries: libcurl
* New executables: git, curl
* Package updates

# v8

* New libraries: libcurl
* New executables: git, curl
* Package updates

# v7

* New compilers: ghc883, ghc8101
* Dropped compilers: ghc844, ghc881
* Package updates

# v6

* ncurses: Set default TERMINFO_DIRS so the resulting binary is a bit more portable

# v5

* Add ncurses library


test/test.sh
#!/usr/bin/env bash

set -o xtrace
set -o errexit

DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"

image=$1

tmpdir=$(mktemp -d)
cp -r "$DIR/../example" "$tmpdir/"

cd "$tmpdir"

cat "$DIR/test-cabal.sh" |
  docker run \
    --rm -v "$tmpdir"/example:/mnt \
    -i "$image" bash

cd "$tmpdir/example"

stack build \
  --ghc-options ' -static -optl-static -optl-pthread -fPIC' \
  --docker --docker-image "$image" \
  --no-nix


test/test-cabal.sh
#!/usr/bin/env bash

set -o xtrace
set -o errexit

cd /mnt
cabal update
cabal new-build example --enable-executable-static


build.sh
#!/usr/bin/env bash

set -o errexit
set -o nounset

function trace() {
    echo "! $@" >&2
    "$@"
}

cd "$( dirname "${BASH_SOURCE[0]}" )"
VERSION="$(jq -r '.version' config.json)"

jq -c '.flavours[]' config.json | {
declare -a tags

while read params; do
  echo "$params"
  ghc_version="$(
    echo "$params" \
      | jq -r '.GHC_VERSION' \
      | tr -d '.' \
  )"

  args=$(
    echo "$params" \
      | jq -r '
          to_entries
            | map(["--build-arg", "\(.key)=\(.value)"])
            | flatten | join(" ")
        '
  )

  target_tag="utdemir/ghc-musl:$VERSION-ghc$ghc_version"

  tmpdir="$(mktemp -d)"
  image="$(
    docker build \
      -f "Dockerfile" \
      $args \
      "$(mktemp -d)" \
      | tee /dev/stderr \
      | tail -n 1 \
      | grep -Po '(?<= ).{12}$'
  )"
  rmdir "$tmpdir"

  trace docker tag "$image" "$target_tag-dev"
  echo | trace "./test/test.sh" "$target_tag-dev"
  trace docker tag "$image" "$target_tag"

  tags+=("$target_tag")
done

echo "Done, written TAGS."
echo -n >TAGS
for tag in "${tags[@]}"
do
     echo "$tag" >> TAGS
done
}


update-readme.sh
#!/usr/bin/env bash

set -o errexit
set -o nounset
set -o pipefail

read -r -d '' template <<'EOF' || true
<!--

WARNING: This file is auto-generated from ./update-readme.sh, modify
that instead.

-->

# ghc-musl

This repository provides Docker images with GHC compiled with `musl`;
therefore can be used to create fully static Haskell binaries without
`glibc` dependency on any platform which can run Docker (x86_64). Powered
by [ghcup](https://www.haskell.org/ghcup/).

Images come with `ghc` and `cabal` executables alongside with commonly
used libraries and build tools. They can also be used with the `stack`
build tool using its Docker integration.

Here are the latest images currently [published in Docker Hub](https://hub.docker.com/r/utdemir/ghc-musl/tags):

${ALL_TAGS}

## Usage

### cabal-install

Mount the project directory to the container, and use `cabal-install`
with `--enable-executable-static` flag inside the container:

```
$ cd myproject/
$ docker run -itv $PWD:/mnt ${EXAMPLE_TAG}
sh$ cd /mnt
sh$ cabal new-update
sh$ cabal new-build --enable-executable-static
```

You can also set `executable-static` [option](https://cabal.readthedocs.io/en/latest/cabal-project.html#cfg-field-executable-static) on your `cabal.project` file.

### stack

Use below arguments, or set the relevant options on your `stack.yaml`:

```
stack build \
  --ghc-options ' -static -optl-static -optl-pthread -fPIC' \
  --docker --docker-image "${EXAMPLE_TAG}" \
  --no-nix
```

Make sure to pick an image with the GHC version compatible with the
Stackage resolver you are using.

Follow https://github.com/commercialhaskell/stack/issues/3420 for
more details on static compilation using the Stack build tool.

### Example session with GHC

Below shell session shows how to start a pre-compiled docker container
and compile a simple `Hello.hs` as a static executable:

```
$ docker run -itv $PWD:/mnt ${EXAMPLE_TAG}
bash-4.4# cd /mnt/
bash-4.4# cat Hello.hs
main = putStrLn "Hello"
bash-4.4# ghc --make -optl-static -optl-pthread Hello.hs
[1 of 1] Compiling Main             ( Hello.hs, Hello.o )
Linking Hello ...
bash-4.4# ls -al Hello
-rwxr-xr-x 1 root root 1185056 Aug 11 16:55 Hello
bash-4.4# file Hello
Hello: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), statically linked, with debug_info, not stripped
bash-4.4# ldd ./Hello
ldd: ./Hello: Not a valid dynamic program
```

The result can be executed on different Linux systems:

```
$ docker run -itv $PWD:/mnt alpine
# /mnt/Hello
Hello
```

```
$ docker run -itv $PWD:/mnt centos
[root@94eb29dcdfe6 /]# /mnt/Hello
Hello
```

## Development

Feel free to open an issue or send a PR to add a library or support a
newer compiler version.

The build process is orchestrated using [Earthly][]. Once it is installed
(or obtained via "nix-shell -p earthly"), you can use below command to build
and test all images:

```
earthly --allow-privileged +all
```

Note: `--allow-privileged` is only necessary because we use Docker-in-Docker
project to test if the generated images work with `stack`'s Docker support.
You can use the following command to disable Stack tests so that
`--allow-privileged` is not necessary:

```
earthly --build-arg TEST_STACK=0 +all
```

The following command updates `README.md`:

```
earthly --artifact +readme/README.md
```

[Earthly]: https://earthly.dev

## Related

* <https://gitlab.com/neosimsim/docker-builder-images>
* Using Nix: <https://github.com/nh2/static-haskell-nix>
* Not maintained: <https://github.com/fpco/docker-static-haskell>
EOF


cd "$( dirname "${BASH_SOURCE[0]}" )"

tags=( "$@" )
export ALL_TAGS="$(echo "${tags[@]}" | tr ' ' '\n' | sed 's/^/* /g')"
export EXAMPLE_TAG="${tags[0]}"

echo "$template" \
  | envsubst '${ALL_TAGS} ${EXAMPLE_TAG}' \
  > README.md

